csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -116,19 +116,17 @@ private module Cached {
   cached
   DataFlowCallable viableCallable(DataFlowCall call) { result = call.getARuntimeTarget() }
 
-  private int parameterPosition() {
-    result =
-      [
-        -1, any(Parameter p).getPosition(),
-        ImplicitCapturedParameterNodeImpl::getParameterPosition(_)
-      ]
-  }
-
   cached
-  newtype TParameterPosition = MkParameterPosition(int i) { i = parameterPosition() }
+  newtype TParameterPosition =
+    TPositionalParameterPosition(int i) { i = any(Parameter p).getPosition() } or
+    TThisParameterPosition() or
+    TImplicitCapturedParameterPosition(SsaCapturedEntryDefinition def)
 
   cached
-  newtype TArgumentPosition = MkArgumentPosition(int i) { i = parameterPosition() }
+  newtype TArgumentPosition =
+    TPositionalArgumentPosition(int i) { i = any(Parameter p).getPosition() } or
+    TQualifierArgumentPosition() or
+    TImplicitCapturedArgumentPosition(SsaCapturedEntryDefinition def)
```

Comment:
This looks risky! This might introduce virtual-dispatch-sized blowups in any part of the dataflow lib that work with an argument without having joined that to a call target.

Reply:
Or is this a case where the call-callee relation is 1-1?

Reply:
Performance-wise this should be no different from before, where we translated each `SsaCapturedEntryDefinition` into a magic number (using `equivalenceRelation`), and used that magic number as both parameter and argument positions.

Reply:
But now that I think about it, this can probably be made a bit nicer (and smaller) by simply using the variable that is captured as the positions.

Reply:
I guess the bad case requires some sort of convoluted setup where a single call might dispatch to many different lambdas (or other kinds of closures) that each capture some variable, and that this single call somehow provides a value in the corresponding argument-position(s). I don't know if such a construction is even possible - it depends on how these arguments are recognised.